File: Binary Search/Q1. Allocate Books/KashishJuneja101003--Q1-Allocate-Books.py
def is_valid(arr, n, m, max_allowed_pages):
    students, pages = 1, 0
    
    for i in range(n):
        if arr[i] > max_allowed_pages:
            return False
        
        if pages + arr[i] <= max_allowed_pages:
            pages += arr[i]
        else:
            students += 1
            pages = arr[i]
    
    return students <= m

def allocate_books(arr, n, m):
    if m > n:
        return -1  # Number of students can't be greater than the number of books
    
    total_pages = sum(arr)
    start, end = 0, total_pages
    result = -1
    
    while start <= end:
        mid = start + (end - start) // 2
        if is_valid(arr, n, m, mid):
            result = mid
            end = mid - 1
        else:
            start = mid + 1
    
    return result

# Example usage
arr = [12, 34, 67, 90]
n, m = 4, 2  # n -> number of books, m -> number of students
print(allocate_books(arr, n, m))

File: Binary Search/Q1. Allocate Books/KashishJuneja101003--Q1-Allocate-Books.cpp
// Time complexity: O(logR*n)
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

bool isValid(vector<int> arr, int n, int m, int maxAllowedPages){
    int students = 1, pages = 0;
    for(int i=0; i<n; i++){ //O(n)
        if(arr[i] > maxAllowedPages) return false;

        if(pages+arr[i] <= maxAllowedPages){
            pages += arr[i];
        }
        else{
            students++;
            pages = arr[i];
        }
    }
    return (students>m) ? false : true;
}

int allocateBooks(vector<int>& arr, int n, int m){  // O(logR * n)
    if(m>n) return -1; // Number of students can't be greater than no of books
    int sum = 0;
    int result = -1;

    // Calculating sum of array
    for(int i:arr) sum += i; //O(n)

    // Creating an array of range in which the answer may lie
    int start = 0, end = sum;
    while (start <= end)    {   //O(log R) R = Range
        int mid = start + (end-start)/2;
        if(isValid(arr, n, m, mid)){    //O(n)
            result = mid;
            end = mid-1;
        }
        else start = mid+1;
    }
    return result;
}

int main(){
    vector<int> arr = {12, 34, 67, 90};
    int n = 4, m = 2; // n-> no of books, m -> no of students

    cout<<allocateBooks(arr, n, m);
    return 0;
}
